src/libraries/Microsoft.Extensions.Configuration.FileExtensions/ref/Microsoft.Extensions.Configuration.FileExtensions.cs

```diff
@@ -40,8 +40,8 @@ public abstract partial class FileConfigurationSource : Microsoft.Extensions.Con
     public partial class FileLoadExceptionContext
     {
         public FileLoadExceptionContext() { }
-        public System.Exception Exception { get { throw null; } set { } }
+        public System.Exception? Exception { get { throw null; } set { } }
```

Comment:
@stephentoub - I am looking for some advice here. I don't like the user experience with this change, but the problem is in the existing code.

Here we have a class whose main use case is to be created by our own code, and passed to consumer code:

https://github.com/dotnet/runtime/blob/ca24ab040b48f4fc0f1bd5e7d2342f42c8bb5ebb/src/libraries/Microsoft.Extensions.Configuration.FileExtensions/src/FileConfigurationProvider.cs#L139-L144

As you can see, we always set `Exception` and `Provider` to non-null values. And then we invoke a registered delegate to "handle" the exception.

With this change, any consumer would need to check these values for `null` before using them. But honestly, what's the point of an "exception context" that doesn't have an exception set?

I'm wondering for ways we can handle this to make it better. One option I think is:

1. Keep these properties as non-nullable
2. Add a new `ctor` that takes non-nullable Exception and Provider.
3. Mark the existing parameterless ctor as `Obsolete`.

Obviously a change like that should go through API review. But I think it will provide a better experience long term.

Do you have any other ideas for approaches here? Have we hit a case like this before when annotating the rest of the libraries for nullable?

Reply:
Chatting with @stephentoub offline, we discussed the following options:

1. Suppress the warnings we get here and leave the properties as non-nullable. We had a similar case for `InterfaceMapping`, which led to the `CONSIDER special-casing public properties and fields on structs.` section in https://github.com/dotnet/runtime/blob/main/docs/coding-guidelines/api-guidelines/nullability.md. In this case, we only ever expect these properties to be set correctly, so marking them as nullable would harm the 100% correct use cases.
2. In the default constructor, we set these properties to dummy instances to ensure they are never null.
3. Alternatively in the future if we see this type being created in scenarios that aren't controlled by this library, we can follow the above plan to obsolete the default ctor and add a new ctor taking non-nullable values.

I would suggest we follow option (1) above for now.

Reply:
How about creating a constructor the sets both fields then? If they always expected to be set anyway.
Yes, this is a breaking change, but it would convey that those fields must be set (which is the intention here). 

https://github.com/dotnet/runtime/pull/57434/commits/a06b5040633ac1dc56b43c78a562b62562d32316